File: backend/app/api/v1/schemas/market.py
"""Market data Pydantic schemas — request/response models for ``/api/v1/market``.

Mirrored on the frontend by ``frontend/src/types/market.ts``.
"""

from __future__ import annotations

from typing import Dict

from pydantic import BaseModel, Field


class TickerData(BaseModel):
    """Single ticker snapshot returned by Yahoo Finance (proxy)."""

    symbol: str = Field(..., description="Ticker symbol, e.g. SPY")
    price: str = Field(..., description="Last regular‑market price, string‑formatted for client convenience")
    change: str = Field(..., description="Regular‑market price change, string‑formatted")
    changePercent: str = Field(..., description="Regular‑market percent change, string‑formatted")


class TickerEnvelope(BaseModel):
    """Standard success envelope: ``{ success: true, data: { SYM: TickerData, ... } }``."""

    success: bool = True
    data: Dict[str, TickerData]

File: frontend/src/types/market.ts
/**
 * Market data types.
 *
 * Wire shape mirrors ``backend/app/api/v1/schemas/market.py`` (Pydantic).
 * ``TickerData`` adds a client‑side ``timestamp`` that the backend never
 * sends — the service layer stamps it after a successful fetch.
 *
 * Backend resource: GET /api/v1/market/tickers
 * Frontend consumer: src/api/services/ticker.service.ts
 */

/** Exact wire shape returned by `/market/tickers` for one symbol. */
export interface RawTicker {
  symbol: string;
  price: string;
  change: string;
  changePercent: string;
}

/** Exact wire envelope: `{ success: true, data: { SYM: RawTicker } }`. */
export interface TickerPayload {
  success: boolean;
  data: Record<string, RawTicker>;
}

/** Client‑side view: same as RawTicker plus a ``timestamp`` set by the service. */
export interface TickerData extends RawTicker {
  timestamp: Date;
}
